lib/java_buildpack/framework/contrast_security_agent.rb

```diff
@@ -0,0 +1,114 @@
+# Encoding: utf-8
+
+# Cloud Foundry Java Buildpack
+# Copyright 2013-2017 the original author or authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'fileutils'
+require 'open-uri'
+require 'rexml/document'
+require 'java_buildpack/component/versioned_dependency_component'
+require 'java_buildpack/framework'
+require 'base64'
+
+module JavaBuildpack
+  module Framework
+
+    # Encapsulates the functionality for running the Contrast Security Agent support.
+    class ContrastSecurityAgent < JavaBuildpack::Component::VersionedDependencyComponent
+
+      # (see JavaBuildpack::Component::BaseComponent#compile)
+      def compile
+        load_credentials
+        FileUtils.mkdir_p(@droplet.sandbox)
+        @dir = @droplet.sandbox
+        download_jar(boot_class_name)
+        build_contrast_configuration
+        @droplet.copy_resources
+      end
+
+      # (see JavaBuildpack::Component::BaseComponent#release)
+      def release
+        app_name = @application.details['application_name'] || 'ROOT'
+        java_opts = @droplet.java_opts
+        java_opts.add_system_property('contrast.dir', '/tmp')
+        java_opts.add_system_property('contrast.override.appname', app_name)
+        path = java_opts.qualify_path(@droplet.sandbox)
+        java_opts.add_preformatted_options("-javaagent:#{path}/#{boot_class_name}=#{path}/contrast.config")
+      end
+
+      protected
+
+      # (see JavaBuildpack::Component::VersionedDependencyComponent#supports?)
+      def supports?
+        @application.services.one_service?(CONTRAST_FILTER, [TEAMSERVER_URL, USERNAME, API_KEY, SERVICE_KEY])
+      end
+
+      private
+
+      CONTRAST_FILTER = /contrast[-]?security/
+      private_constant :CONTRAST_FILTER
+
+      TEAMSERVER_URL = 'teamserver_url'.freeze
+      USERNAME = 'username'.freeze
+      API_KEY = 'api_key'.freeze
+      SERVICE_KEY = 'service_key'.freeze
+
+      PLUGIN_PACKAGE = 'com.aspectsecurity.contrast.runtime.agent.plugins.'.freeze
+
+      def load_credentials
+        credentials = @application.services.find_service(CONTRAST_FILTER)['credentials']
+        @teamserver_url = credentials[TEAMSERVER_URL]
+        @username = credentials[USERNAME]
+        @api_key = credentials[API_KEY]
+        @service_key = credentials[SERVICE_KEY]
+      end
+
+      def boot_class_name
+        version = @version.to_s.split('_')[0]
+        "contrast-engine-#{version}.jar"
+      end
+
+      def build_contrast_configuration
```

Comment:
I'll leave this one to you to decide, but making a real live XML document might not be the easiest or clearest way to do this.  Another option is to use extended literals and just write clear text that you know to be valid XML.  [An example](https://github.com/cloudfoundry/java-buildpack/blob/master/lib/java_buildpack/framework/luna_security_provider.rb#L146-L163).  It's up to you though, and if you think this is easier, I'm happy to leave it as-is.